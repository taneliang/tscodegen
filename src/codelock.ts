import crypto from "crypto";
import {
  removeFileDocblock,
  getFileDocblock,
  prependFileDocblock,
} from "./sections/docblock";
import { emptyManualSections } from "./sections/manual";

interface CodelockInfo {
  hash: string;
  manualSectionsAllowed: boolean;
}

/**
 * Get codelock information from a locked source file.
 *
 * @param lockedCode Code in source file, prepended with codelock file docblock.
 */
export function getCodelockInfo(lockedCode: string): CodelockInfo | undefined {
  // TODO: get docblock, and if it exists, retrieve hash
  const docblock = getFileDocblock(lockedCode);
  if (!docblock) {
    return undefined;
  }

  const docblockLines = docblock.split("\n");
  if (docblockLines.length === 0) {
    return undefined;
  }

  // Expect codelock info to be on the last line.
  const lockline = docblockLines[docblockLines.length - 1];

  const editableMatchGroups = /^@generated-editable Codelock<<(?<hash>\S+?)>>$/.exec(
    lockline
  )?.groups;
  if (editableMatchGroups) {
    return {
      hash: editableMatchGroups.hash,
      manualSectionsAllowed: true,
    };
  }

  const uneditableMatchGroups = /^@generated Codelock<<(?<hash>\S+?)>>$/.exec(
    lockline
  )?.groups;
  if (uneditableMatchGroups) {
    return {
      hash: uneditableMatchGroups.hash,
      manualSectionsAllowed: false,
    };
  }

  return undefined;
}

/**
 * Computes the codelock hash for a source file. Will not unlock if locked.
 *
 * @param code Code to be locked.
 * @param shouldEmptyManualSections Whether manual sections should be emptied.
 * Should = whether the file is allowed to have manual sections.
 * @returns Lock hash for `code`.
 */
function computeHash(code: string, shouldEmptyManualSections: boolean): string {
  const hashableCode = (shouldEmptyManualSections
    ? emptyManualSections(code)
    : code
  ).trim();
  return crypto
    .createHash("shake128", { outputLength: 24 })
    .update(hashableCode)
    .digest("base64");
}

/**
 * Computes the hash for the source code, and prepends a codelock file
 * docblock. Will **NOT** unlock if locked, in case the first thing in the file
 * is an unrelated docblock.
 *
 * @param code Code to be locked.
 * @param manualSectionsAllowed Whether generated code can contain manual sections.
 * @returns Locked code, i.e. code with prepended codelock file docblock.
 */
export function lockCode(code: string, manualSectionsAllowed: boolean): string {
  const hash = computeHash(code, manualSectionsAllowed);

  let docblockContent;

  if (manualSectionsAllowed) {
    docblockContent = `This file is generated with manually editable sections. Only make
modifications between BEGIN MANUAL SECTION and END MANUAL SECTION
designators.

@generated-editable Codelock<<${hash}>>`;
  } else {
    docblockContent = `This file is generated. Do not modify it manually.

@generated Codelock<<${hash}>>`;
  }

  return prependFileDocblock(code, docblockContent);
}

/**
 * Verify that the codelock in the source file is valid.
 *
 * @param lockedCode Locked code to be verified.
 * @returns `true` if lock is found and verified. `false` otherwise.
 */
export function verifyLock(lockedCode: string): boolean {
  const codeblockInfo = getCodelockInfo(lockedCode);
  if (!codeblockInfo) {
    return false;
  }
  return (
    codeblockInfo.hash ===
    computeHash(
      removeFileDocblock(lockedCode),
      codeblockInfo.manualSectionsAllowed
    )
  );
}
