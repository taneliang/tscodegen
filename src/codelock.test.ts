import { getCodelockInfo, lockCode, verifyLock } from "./codelock";

const codeWithoutDocblock = `
interface CodelockInfo {
  hash: string;
  manualSectionsAllowed: boolean;
  /* BEGIN MANUAL SECTION custom_fields */
  /* END MANUAL SECTION */
}
`.trim();

describe(getCodelockInfo, () => {
  test("should return undefined if code has no docblock", () => {
    expect(getCodelockInfo("")).toBeUndefined();
    expect(getCodelockInfo(codeWithoutDocblock)).toBeUndefined();
  });

  test("should return undefined if file has some non-codelock docblock", () => {
    const code = `
/**
 * I swear, this file is generated. Believe me. Please don't edit!
 *
 * @totally-generated Codelock<<fakehash>>
 */

interface CodelockInfo {
  hash: string;
  manualSectionsAllowed: boolean;
}
`;
    expect(getCodelockInfo(code)).toBeUndefined();
  });

  test("should return codelock info for editable files", () => {
    const lockedEditableCode = `
/**
 * This file is generated with manually editable sections. Only make
 * modifications between BEGIN MANUAL SECTION and END MANUAL SECTION
 * designators.
 *
 * @generated-editable Codelock<<somehash>>
 */

interface CodelockInfo {
  hash: string;
  manualSectionsAllowed: boolean;
}

/**
 * Get codelock information from a locked source file.
 *
 * @param lockedCode Code in source file, prepended with codelock file docblock.
 */
export function getCodelockInfo(lockedCode: string): CodelockInfo | undefined {
  /* BEGIN MANUAL SECTION manual */
  console.log("some manual code");
  /* END MANUAL SECTION */
}
`.trim();

    expect(getCodelockInfo(lockedEditableCode)).toEqual({
      hash: "somehash",
      manualSectionsAllowed: true,
    });
  });

  test("should return codelock info for uneditable files", () => {
    const lockedUneditableCode = `
/**
 * This file is generated. Do not modify it manually.
 *
 * @generated Codelock<<somehash>>
 */

interface CodelockInfo {
  hash: string;
  manualSectionsAllowed: boolean;
}

/**
 * Get codelock information from a locked source file.
 *
 * @param lockedCode Code in source file, prepended with codelock file docblock.
 */
export function getCodelockInfo(lockedCode: string): CodelockInfo | undefined {}
`.trim();

    expect(getCodelockInfo(lockedUneditableCode)).toEqual({
      hash: "somehash",
      manualSectionsAllowed: false,
    });
  });
});

describe(lockCode, () => {
  describe("editable files", () => {
    test("should prepend file docblock with consistent hash", () => {
      const lockedCode = lockCode(codeWithoutDocblock, true);
      expect(lockedCode).toContain("@generated-editable Codelock<<");
      expect(lockedCode).toMatchSnapshot();
    });
  });

  describe("uneditable files", () => {
    test("should prepend file docblock with consistent hash", () => {
      const lockedCode = lockCode(codeWithoutDocblock, false);
      expect(lockedCode).toContain("@generated Codelock<<");
      expect(lockedCode).toMatchSnapshot();
    });
  });
});

describe(verifyLock, () => {
  test("should return false if file does not have docblock", () => {
    expect(verifyLock("")).toBe(false);
    expect(verifyLock(codeWithoutDocblock)).toBe(false);
  });

  test("should return true if file has not been modified after locking", () => {
    expect(verifyLock(lockCode(codeWithoutDocblock, true))).toBe(true);
    expect(verifyLock(lockCode(codeWithoutDocblock, false))).toBe(true);
  });

  test("should return false if file has been modified outside manual sections after locking", () => {
    expect(verifyLock(lockCode(codeWithoutDocblock, true) + "\naaaaaa")).toBe(
      false
    );
    expect(verifyLock(lockCode(codeWithoutDocblock, false) + "\naaaaaa")).toBe(
      false
    );
    expect(
      verifyLock(
        lockCode(codeWithoutDocblock, true).replace(
          "MANUAL SECTION custom_fields",
          "MANUAL SECTION a_bit_too_custom"
        )
      )
    ).toBe(false);
  });

  describe("editable files", () => {
    test("should return true if file has been modified in manual sections after locking", () => {
      expect(
        verifyLock(
          lockCode(codeWithoutDocblock, true).replace(
            "/* BEGIN MANUAL SECTION custom_fields */\n",
            "/* BEGIN MANUAL SECTION custom_fields */\nconsole.log('Custom code');\n"
          )
        )
      ).toBe(true);
    });
  });

  describe("uneditable files", () => {
    test('should return false if file has been modified in "manual" sections after locking', () => {
      expect(
        verifyLock(
          lockCode(codeWithoutDocblock, false).replace(
            "/* BEGIN MANUAL SECTION custom_fields */\n",
            "/* BEGIN MANUAL SECTION custom_fields */\nconsole.log('Custom code in uneditable file');\n"
          )
        )
      ).toBe(false);
    });
  });
});

describe("reversibility", () => {
  test("lock -> getCodelockInfo should return the same manualSectionsAllowed", () => {
    expect(getCodelockInfo(lockCode(codeWithoutDocblock, true))).toMatchObject({
      manualSectionsAllowed: true,
    });
    expect(getCodelockInfo(lockCode(codeWithoutDocblock, false))).toMatchObject(
      {
        manualSectionsAllowed: false,
      }
    );
  });
});
