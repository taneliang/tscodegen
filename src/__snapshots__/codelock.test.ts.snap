// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lockCode editable files should interpolate provided content 1`] = `
"/**
 * This file is generated with manually editable sections. Only make
 * modifications between BEGIN MANUAL SECTION and END MANUAL SECTION
 * designators.
 * CUSTOM LINE 1
 * CUSTOM LINE 2
 * @generated-editable Codelock<<VZ4MZwQh3kgACluPomV+XBbiKqmafUfq>>
 */

interface CodelockInfo {
  hash: string;
  manualSectionsAllowed: boolean;
  /* BEGIN MANUAL SECTION custom_fields */
  /* END MANUAL SECTION */
}"
`;

exports[`lockCode editable files should prepend file docblock with consistent hash 1`] = `
"/**
 * This file is generated with manually editable sections. Only make
 * modifications between BEGIN MANUAL SECTION and END MANUAL SECTION
 * designators.
 *
 * @generated-editable Codelock<<VZ4MZwQh3kgACluPomV+XBbiKqmafUfq>>
 */

interface CodelockInfo {
  hash: string;
  manualSectionsAllowed: boolean;
  /* BEGIN MANUAL SECTION custom_fields */
  /* END MANUAL SECTION */
}"
`;

exports[`lockCode uneditable files should interpolate provided content 1`] = `
"/**
 * This file is generated. Do not modify it manually.
 * CUSTOM LINE 1
 * CUSTOM LINE 2
 * @generated Codelock<<SZOHkis46G+poAie7wLZwJjvgG/PEyVm>>
 */

interface CodelockInfo {
  hash: string;
  manualSectionsAllowed: boolean;
  /* BEGIN MANUAL SECTION custom_fields */
  /* END MANUAL SECTION */
}"
`;

exports[`lockCode uneditable files should prepend file docblock with consistent hash 1`] = `
"/**
 * This file is generated. Do not modify it manually.
 *
 * @generated Codelock<<SZOHkis46G+poAie7wLZwJjvgG/PEyVm>>
 */

interface CodelockInfo {
  hash: string;
  manualSectionsAllowed: boolean;
  /* BEGIN MANUAL SECTION custom_fields */
  /* END MANUAL SECTION */
}"
`;
